### PR TITLE
API: Add numpy.testing.overrides to aid testing of custom array containers

### LINKED ISSUE #15544: DOC: Listing of all numpy functions that support __array_function__
Is there a listing of all numpy functions one would need to wrap to fully support all of the API that is wrappable via `__array_function__`? This seems like something that could be generated by introspection on the numpy source code and included in the numpy documentation, as far as I can see it doesn't exist.

Having such a document to look at would make it much easier for downstream projects to add `__array_function__` support. See https://github.com/yt-project/unyt/issues/139 for more details of my usage.

### PR DESCRIPTION
**Pull request by ngoldbaum**
Closes #15544

It wasn't clear to me how to test this, suggestions along those lines are appreciated.

### CODE REVIEW

File: numpy/core/overrides.py
Comment (hameerabbasi) [inline numpy/core/overrides.py:11]:
```suggestion
ARRAY_FUNCTIONS = set()
```

File: numpy/testing/overrides.py
Comment (hameerabbasi) [inline numpy/testing/overrides.py:11]:
```suggestion
ufuncs = set(_umath_mod.__dict__.values())
```

**Review by hameerabbasi — CHANGES_REQUESTED** [review]
It may just be me, but I don't see a value in having these ordered, but I do see value in fast membership tests like the ones you do.

File: numpy/core/overrides.py
Comment (hameerabbasi) [inline numpy/core/overrides.py:213]:
```suggestion
        ARRAY_FUNCTIONS.add(public_api)
```

**Review by seberg — COMMENTED** [review]
Thanks, especially for adding new docs also!  I was wondering if we care about the sets being mutable, but this seems a grown-up corner...

I just realized that things are a bit funny: technically `from numpy.lib import recfunctions` requires explicit import for its functions to show up here (I suspect that is the only part in NumPy where this happens, but am not sure).

Should we be pragmatic and ignore that problem, considering that this is targeted to downstream library authors?
The only idea I have right now would be tying it into a more involved test.

For testing, the only thing I can think right now would be to make a bigger public API test with a list of all API symbols (in the "public" namespaces).
That would be more generally helpful (no accidental API additions to the main namespace), and then could include the information for ufunc/`__array_function__`.
But: That is a bit more involved, so I think I am OK if we skip it.  It would be good to post the full output for a brief manual check though.

File: numpy/testing/overrides.py
Comment (ngoldbaum) [reply numpy/testing/overrides.py:12]: I didn't know there were ufuncs in there! This set should only include things in the public numpy API, so no I don't think they should be included. I've also added a note to the docs that `ufuncs` and `array_functions` only include things in the public numpy API.

**Comment by ngoldbaum** [comment]
> I just realized that things are a bit funny: technically from numpy.lib import recfunctions requires explicit import for its functions to show up here (I suspect that is the only part in NumPy where this happens, but am not sure).

Good call! I think the most straightforward fix is to make it to mirror that and make users explicitly import `numpy.testing.overrides`. That way it's safe to import stuff inside of `numpy.testing.overrides` that importing the top-level `numpy` module doesn't ensure is defined. I think users who want to use this API would prefer it if `array_functions` contained *all* of the overridable functions irrespective of how their imports are setup.

> For testing, the only thing I can think right now would be to make a bigger public API test with a list of all API symbols (in the "public" namespaces).

Yeah, that's a good point. I'll open an issue to track that idea once this is merged, but I'd prefer not to add that test as part of this if that's OK.

File: numpy/testing/overrides.py
Comment (seberg) [reply numpy/testing/overrides.py:12]: Right, they are probably wrapped by `__array_function__` anyway (in functions that support it).  One day it would be nice to figure out the part of falling back to NumPy for part of the array-function API (it doesn't seem impossible, but needs someone dedicated to do some careful scouting, I suspect).
Only then it would probably even start to make sense to bless some ufuncs to explicitly show up here (e.g. the internal clip ufunc).

**Comment by seberg** [comment]
Sorry, I mean just a function, like `get_numpy_ufuncs()` (or a name without get also) rather than an attribute.

**Comment by ngoldbaum** [comment]
Added the getters

File: numpy/testing/overrides.py
Comment (seberg) [inline numpy/testing/overrides.py:24]: This is for array-functions.  It should not matter for ufuncs.

**Review by seberg — APPROVED** [review]
Thanks LGTM, modulo moving the manual import.  Lets just see if @shoyer has any further opinions.

**Comment by seberg** [comment]
Well, silence is agreement (unfortunately), so putting this in.  This adds:
* `allows_array_function_override`
* `allows_array_ufunc_override`
* `get_overridable_numpy_ufuncs`
* `get_overridable_numpy_array_functions`

and the `numpy.testing.overrides` module to host them.  @shoyer if you don't like this or have thoughts, please comment and we can follow up.

Thanks Nathan.

**Comment by shoyer** [comment]
This looks good to me, too. Thanks Nathan!

**Pull request merged**
